### PR TITLE
[SPARK-49699][SS] Disable PruneFilters for streaming workloads

### PR DESCRIPTION
**Pull request by n-young-db**
### What changes were proposed in this pull request?

The PR proposes to disable PruneFilters if the predicate of the filter is evaluated to `null` / `false` and the filter (and subtree) is streaming.

### Why are the changes needed?

PruneFilters replaces the `null` / `false` filter with an empty relation, which means the subtree of the filter is also lost. The optimization does not care about whichever operator is in the subtree, hence some important operators like stateful operator, watermark node, observe node could be lost.

The filter could be evaluated to `null` / `false` selectively among microbatches in various reasons (one simple example is the modification of the query during restart), which means stateful operator might not be available for batch N and be available for batch N + 1. For this case, streaming query will fail as batch N + 1 cannot load the state from batch N, and it's not recoverable in most cases.

See new tests in StreamingQueryOptimizationCorrectnessSuite for details.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

UT.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3832]: can we also add version?

**Comment by HyukjinKwon** [comment]
> If PruneFilters triggers, we can lose a stateful operator and fail a query.

Could we explain this better? Why and how the stateful operator is lost, and how does it fail?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (HeartSaVioR) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3835]: @n-young-db 
Spark 3.5.2 is already released. For safety, this would be better to be set as 4.0.0, as we don't justify the necessity of backport.

**Review by HeartSaVioR — COMMENTED** [review]
The test coverage seems to be a bit weak, but I can work with @n-young-db to come up with closer-to-e2e tests.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (HeartSaVioR) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3835]:
```suggestion
      .version("4.0.0")
```

**Review by HeartSaVioR — APPROVED** [review]
+1 pending CI

@HyukjinKwon @cloud-fan 
I pushed my own change for `sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryOptimizationCorrectnessSuite` as it's lot easier for me to reason about e2e tests.

Since there is few active committers around SS so not sure who could help review my change, but I hope I could get review from you. (I'll just go ahead if you don't feel like able to review the change.)

Thanks in advance!

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryOptimizationCorrectnessSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryOptimizationCorrectnessSuite.scala:585]: Without SPARK-49699, the (stateful) streaming aggregation operator is lost.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryOptimizationCorrectnessSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryOptimizationCorrectnessSuite.scala:541]: Without [SPARK-49699](https://issues.apache.org/jira/browse/SPARK-49699), we couldn't find the observed metrics named "observation".

**Comment by HeartSaVioR** [comment]
OK, I've waited for 1 day and no feedback. Let me handle further input as post review; I approve the part @n-young-db is written. My code is only for testing.

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master/3.5. (Let's merge this to 3.5 as well as I've been merging the streaming & QO fix to 3.5.)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3831]: oops the name shouldn't have databricks

Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3831]: Ouch... Let's fix it. cc. @n-young-db

**Pull request closed without merging**
